cmd/acra-reader/server.go

```diff
@@ -171,59 +171,67 @@ func (server *ReaderServer) processGRPCConnection(parentContext context.Context,
 	logger := logging.GetLoggerFromContext(parentContext)
-	logger.Debugln("http handler")
+	logger.Debugln("HTTP handler")
 
 	reader := bufio.NewReader(connection)
 	request, err := http.ReadRequest(reader)
-	response := http.Response{
-		Status:        "200 OK",
-		StatusCode:    200,
-		Proto:         "HTTP/1.1",
-		ProtoMajor:    1,
-		ProtoMinor:    1,
-		Request:       request,
-		ContentLength: -1,
-		Header:        http.Header{},
-	}
+
+	// TODO: handle keep alive
 
 	if err != nil {
 		logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderCantHandleHTTPRequest).
 			Warningln("Got new HTTP request, but can't read it")
-		response.StatusCode = http.StatusBadRequest
-		closeConnectionAndSendResponse(logger, response, connection)
+		server.closeConnectionAndSendResponse(logger, emptyResponseWithStatus(request, http.StatusBadRequest), connection)
 		return
 	}
 
-	log.Debugf("Incoming API request to %v", request.URL.Path)
+	response := server.parseRequestPrepareResponse(logger, request, clientId)
+	server.closeConnectionAndSendResponse(logger, response, connection)
+}
+
+
+func (server *ReaderServer) closeConnectionAndSendResponse(logger *log.Entry, response *http.Response, connection net.Conn) {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.BigEndian, response)
+
+	if err != nil {
+		logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderCantReturnResponse).
+			Warningln("Can't write response to HTTP request")
+	} else {
+		connection.Write(buf.Bytes())
+	}
+
+	connection.Close()
+}
 
-	// TODO: handle keep alive
+
+func (server *ReaderServer) parseRequestPrepareResponse(logger *log.Entry, request *http.Request, clientId []byte) *http.Response {
+	if request == nil || request.URL == nil {
+		return emptyResponseWithStatus(request, http.StatusBadRequest)
+	}
+
+	log.Debugf("Incoming API request to %v", request.URL.Path)
 
 	if request.Method != http.MethodPost {
-		logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderMethodNotAllowed).
-			Warningf("HTTP method is not allowed, expected /POST, got %s", request.Method)
-		response.StatusCode = http.StatusMethodNotAllowed
-		closeConnectionAndSendResponse(logger, response, connection)
-		return
+		logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderMethodNotAllowed).
+			Warningf("HTTP method is not allowed, expected POST, got %s", request.Method)
+		return emptyResponseWithStatus(request, http.StatusMethodNotAllowed)
 	}
 
 	// /v1/decrypt
 	// /, v1, decrypt
 	pathParts := strings.Split(request.URL.Path, string(os.PathSeparator))
 	if len(pathParts) != 3 {
-		logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderMalformedURL).
+		logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorReaderMalformedURL).
 			Warningf("Malformed URL, expected /<version>/<endpoint>, got %s", request.URL.Path)
```

Comment:
can we place this explanation to response body to simplify understanding of errors by clients?